doc/style-guide-testing.md

```diff
@@ -0,0 +1,467 @@
+
+
+### Every Path, Success And Failure
+
+We aim for 100% code coverage.  There's no real good reason for us to avoid it.  When there is a problem testing code that prevents us from achieving 100% coverage then we should correct the problem instead of skipping tests.
+
+Every true and every false condition should be tested.  Every promise should be resolved successfully and also should be rejected.  When considering options, make sure to handle the error scenarios as well as every type of success scenario.
+
+This may seem to contradict the "Test Behavior" section because we must be able to inspect the code to see every possible logical path that would be taken.  There is no conflict - each logical path results in outward behavior that the user might encounter.  Even if the code is later refactored to eliminate many logical paths, the tests should still pass and there would be no need to delete any of the tests that now follow a redundant path.
+
+
+Sample Test
+-----------
+
+This is a sample test for a directive in Angular.  Let's pretend our directive will simply set some property on an isolated scope to the value passed into the directive.  There are many things wrong with this test.
+
+    /*global angular, describe, expect, inject, it, module*/
+    describe('myCustomDirective', function () {
+        var $scope;
+
+        beforeEach(module('app'));
+        it('some property on scope', inject(function ($compile, $rootScope) {
+            $scope = $rootScope.$new();
+            $compile(angular.element('<div my-custom-directive="\'test\'"></div>'))($scope);
+            $scope.$digest();
+            expect($scope.some).toEqual('test');
+        });
+        it('some property on scope', inject(function ($compile, $rootScope) {
+            $scope = $rootScope.$new();
+            $scope.test = {
+                thisIsATest: true,
+                unknownThingsAreAllowed: 'Yeah, I guess some are allowed',
+                yWords: [
+                    'yawn',
+                    'yellow',
+                    'yodel',
+                    'yuck'
+                ]
+            };
+            $compile(angular.element('<div my-custom-directive="test"></div>'))($scope);
+            $scope.$digest();
+            expect($scope.some).toEqual({
+                thisIsATest: true,
+                unknownThingsAreAllowed: 'Yeah, I guess some are allowed',
+                yWords: [
+                    'yawn',
+                    'yellow',
+                    'yodel',
+                    'yuck'
+                ]
+            });
+        });
+    });
+
+What is wrong?
+
+* The data passed in as $scope.test does not need to have a lot of data.  What the directive accomplishes is merely setting the directive's `$scope.some` to the value itself.
+
+* Since we know the object that `$scope.some` should be, we should use `.toBe()` instead of `.toEqual()`.
```

Comment:
I don't think you can use `.toBe()` when comparing objects since the instance of those objects are different, hence when comparing objects you need to use `.toEqual()`.  That's they way I've always had to do it when using Jasmine.